<a href="https://colab.research.google.com/github/sriva-e/Bombora-Data-Science-exam-/blob/main/ds_interview_exam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div class="clearfix" style="padding: 10px; padding-left: 0px">
<a href="http://bombora.com"><img src="https://app.box.com/shared/static/e0j9v1xjmubit0inthhgv3llwnoansjp.png" width="200px" class="pull-right" style="display: inline-block; margin: 5px; vertical-align: middle;"></a>
<h1> Bombora Data Science: <br> *Interview Exam* </h1>
</div>

<img width="200px" src="https://app.box.com/shared/static/15slg1mvjd1zldbg3xkj9picjkmhzpa5.png">

---
# Welcome

Welcome! This notebook contains interview assessment questions referenced in the *Instructions* section in the `README.md`—please read that first, *before* attempting to answer questions here.

<div class="alert alert-info" role="alert" style="margin: 10px">
<p style="font-weight:bold">ADVICE</p>
<p>*Do not* read these questions, and panic, *before* reading the instructions in `README.md`.</p>
</div>

## Have fun!

Regardless of outcome, getting to know you is important. Give it your best shot and we look forward to reading your results and following up! Remember, even if you do not completely finish a question from each section, its fine.

# Exam Questions

## 1. Algo + Data Structures
#### Please limit yourself to a half hour for this section

### Q 1.1: Fibionacci
![fib image](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/200px-Fibonacci_spiral_34.svg.png)

#### Q 1.1.1
Given $n$ where $n \in \mathbb{N}$ (i.e., $n$ is an integer and $n > 0$), write a function `fibonacci(n)` that computes the Fibonacci number $F_n$, where $F_n$ is defined by the recurrence relation:

$$ F_n = F_{n-1} + F_{n-2}$$

with initial conditions of:

$$ F_1 = 1,  F_2 = 1$$

In [ ]:
def fibonacci(n):
    
    if (n == 0 or n == 1):
        
  
        return(n)
  
    else:
        
  
        return (fibonacci(n - 1) + fibonacci(n - 2))
  

In [ ]:
fibonacci(9)

34

#### Q 1.1.2
What's the complexity of your implementation?

In the above code T(n)=T(n-1)+T(n-2), We are taking top to bottom approach in this function , So it will have same complexity as Fibonacci , as Fibonacchi grown exponentially , so will the time complexity of this  function. 
The time complexity of this code is O(2^n) or exponential . 


#### Q 1.1.3
Consider an alternative implementation to compute Fibonacci number $F_n$ and write a new function, `fibonacci2(n)`.

In [ ]:
def fibonacci2(n):
     
    
    fibo = [0, 1]
     
     
    for i in range(2, n+1):
        fibo.append(fibo[i-1] + fibo[i-2])
        
    return fibo[n]

In [ ]:
fibonacci2(9)

34

#### Q 1.1.4
What's the complexity of your implementation?

The time complexity of this code is  O(n) , as we are looping till n times and the computations inside the loop are linear. In this we are taking a bottom to up approach while calculating the nth fibonacci number, although this code takes more memory but the compuatation will be faster.  

#### Q 1.1.5
(Optional) What are some examples of optimizations that could improve computational performance?


We can write another function which calculates the Power of the Matrix  \begin{bmatrix} \ 1 & \ 1 \\ \ 1 & \ 0 \end{bmatrix} **n** times,
this will give us the **(n+1)th** fibonacci number in the resultant matrix , the first entry in first row and first column will be this number.
This will improve the time complexity of our code, as Time complexity for this will be O(log n)

### Q 1.2: Linked List
![ll img](https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Singly-linked-list.svg/500px-Singly-linked-list.svg.png)

#### Q 1.2.1
Consider a [singly linked list](https://en.wikipedia.org/wiki/Linked_list), $L$. Write a function `is_palindrome(L)` that detects if $L$ is a [palindrome](https://en.wikipedia.org/wiki/Palindrome), by returning a bool, `True` or `False`.


#### Q 1.2.2
What is the complexity of your implementation?

#### Q 1.2.3
Consider an alternative implementation to detect if L is a palindrome and write a new function, `is_palindrome2(L)`.

#### Q 1.2.4
What's the complexity of this implementation?


#### Q 1.2.5 
(Optional) What are some examples of optimizations that could improve computational performance?


----
## 2. Prob + Stats
#### Please limit yourself to a half hour for this section

### Q 2.1: Making a 6-side die roll a 7?

Using a single 6-side die, how can you generate a random number between 1 - 7?

Any 6 sided die will only give outcomes from 1-6, To divide it into 7 partitions, I will consider that we are rolling this die twice, now we have 36 outcomes but that also cannot be partioned into 7. So I will reject one of the outcomes , For example if I get (6,6) I will not consider it and roll the second die again to get a different outcome. Now these 35 outcomes can be divided into 5 partitions. So if I have an array of 35 elements where I am storing all these outcomes , I will divide the index by 7 and add 1 , that's how I can get a random number between 1-7 (Here I am assuming array indexing starts from 1)

The given die is 6 sided , The 

### Q 2.2: Is normality uniform?

<img src=https://rednaxela1618.files.wordpress.com/2014/06/uniformnormal.png width="480">


Given draws from a normal distribution with known parameters, how can you simulate draws from a uniform distribution?

### Q 2.3: Uber vs. Lyft

![uber vs lyft](http://usiaffinity.typepad.com/.a/6a01347fc1cb08970c01bb0876bcbe970d-pi)

You request 2 UberX’s and 3 Lyfts. If the time that each takes to reach you is IID (Independent and identically distributed), what is the probability that all the Lyfts arrive first? What is the probability that all the UberX’s arrive first?

All Lyft's first

probability that the first car is Lyft = 3/5
probability that the second car is Lyft = 2/4
probability that the third car is Lyft = 1/3 Therefore, probability that all the Lyfts arrive first is (3/5) * (2/4) * (1/3) = 1/10

All Uber's first
probability that the first car is Uber=2/5
probability that the second car is Uber=1/4
Therefore, probability that all the Ubers arrive first is (2/5)*(1/4)=1/10

The probability that a lyft arrives or an Uber arrives is same :) !


--------
## 3. NLP Algorithm Development
#### Please limit yourself to three hours for this section

Please provide all code and results. Use markdown cells to explain your approach. If there are pieces you couldn’t get to, or approaches you did not have time to explore, then please document your thoughts.

### Q 3.1 Build a Multi-Class Classifier

#### Overview

This exercise is a representation of the problem sets that are critical to our platform and success. You are asked to create a multi-class classification problem using the “option_1.csv” dataset. There is a list of 5K reviews for several product categories. Create a classification model to classify the review text into each category.

#### Data
“Text” – A review given by a customer for a purchased item

“Cat” – A category for the item being reviewed

In [ ]:
import pandas as pd
df = pd.read_csv('https://github.com/bomboradata/datascience-exam/raw/master/option_1.csv')
df

Text  \
0     Hi! I'm Martine Redman and I created this puzz...   
1     120 colors? I say 120 sticks of fun! And a fre...   
2     What a wonderful creation. Art, in our house, ...   
3     Meowth is the coolest, lots of play value. Gre...   
4     You can give a MagnaDoodle to any kid who's ab...   
...                                                 ...   
4995  I am always looking for foods that have not be...   
4996  Critical item for our household which includes...   
4997  I ordered these based on the positive reviews ...   
4998  I was a little leery having to buy a box with ...   
4999  I purchased this for making fudge. I made 2 ba...   

                          Cat  
0                     puzzles  
1                 arts crafts  
2                 arts crafts  
3          action toy figures  
4                 arts crafts  
...                       ...  
4995               snack food  
4996                baby food  
4997               snack food  
4998               snack food  
4999  cooking baking supplies  

[5000 rows x 2 columns]

In [ ]:
#cleaning and preprocessing the data 
import re
from nltk.corpus import stopwords
from wordcloud import WordCloud, STOPWORDS
import spacy

In [ ]:
df['new_Text'] = df['Text'].apply(lambda x: " ".join(x.lower() for x in x.split()))
df['new_Text'].head()

0    hi! i'm martine redman and i created this puzz...
1    120 colors? i say 120 sticks of fun! and a fre...
2    what a wonderful creation. art, in our house, ...
3    meowth is the coolest, lots of play value. gre...
4    you can give a magnadoodle to any kid who's ab...
Name: new_Text, dtype: object

In [ ]:
df['new_Text'] = df['new_Text'].str.replace('[^\w\s]','')
df['new_Text'].head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  """Entry point for launching an IPython kernel.


0    hi im martine redman and i created this puzzle...
1    120 colors i say 120 sticks of fun and a free ...
2    what a wonderful creation art in our house com...
3    meowth is the coolest lots of play value great...
4    you can give a magnadoodle to any kid whos abl...
Name: new_Text, dtype: object

In [ ]:
import nltk
nltk.download('stopwords')
stop = stopwords.words('english')
df['new_Text'] = df['new_Text'].apply(lambda x: " ".join(x for x in x.split() if x not in stop))
df.head(20)

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


Text                    Cat  \
0   Hi! I'm Martine Redman and I created this puzz...                puzzles   
1   120 colors? I say 120 sticks of fun! And a fre...            arts crafts   
2   What a wonderful creation. Art, in our house, ...            arts crafts   
3   Meowth is the coolest, lots of play value. Gre...     action toy figures   
4   You can give a MagnaDoodle to any kid who's ab...            arts crafts   
5   Mitchum Clear Gel Anti-Perspirant is an effect...          personal care   
6   I loved the flavor of the bars!!! I really did...     nutrition wellness   
7   My 5 year old daughter loved following the lif...     learning education   
8   On my way to a great vacation in the Pocono's ...   electronics for kids   
9   These things are absolutely marvelous! Serious...     household supplies   
10  I CONFESS I AM AN ADDICT!! Recently I recieved...   electronics for kids   
11  If you are one of those people that collects t...  stuffed animals plush   
12  Peace is a beautiful multicolored bear. No two...  stuffed animals plush   
13  I have played checkers for over a year now, an...   electronics for kids   
14  I've used these batteries my entire life in ev...     household supplies   
15  My kids have alot of fun with this spongy clay...            arts crafts   
16  I liked the Breath Right Nasal Strips. They ke...            health care   
17  This shaver was much better than others I have...          personal care   
18  I play for a college soccer team, and on one o...   electronics for kids   
19  FLIMSY FLIMSY FLIMSY! FOR THE PRICE YOU THINK ...            arts crafts   

                                             new_Text  
0   hi im martine redman created puzzle briarpatch...  
1   120 colors say 120 sticks fun free sharpener s...  
2   wonderful creation art house comes many colors...  
3   meowth coolest lots play value great use kids ...  
4   give magnadoodle kid whos able grasp stylus ba...  
5   mitchum clear gel antiperspirant effective deo...  
6   loved flavor bars really didnt expect good nut...  
7   5 year old daughter loved following life cycle...  
8   way great vacation poconos son family niita to...  
9   things absolutely marvelous seriously folks li...  
10  confess addict recently recieved slingo gift b...  
11  one people collects things means get thrill am...  
12  peace beautiful multicolored bear two alike so...  
13  played checkers year im quite serious game hop...  
14  ive used batteries entire life everything smok...  
15  kids alot fun spongy clay love mixing new colo...  
16  liked breath right nasal strips kept nostrils ...  
17  shaver much better others tried past didnt lea...  
18  play college soccer team one away games one gi...  
19  flimsy flimsy flimsy price think could make li...

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(sublinear_tf=True, min_df=5, norm='l2', encoding='latin-1', ngram_range=(1, 2), stop_words='english')
features = tfidf.fit_transform(df.new_Text).toarray()
labels = df.Cat

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.naive_bayes import MultinomialNB
X_train, X_test, y_train, y_test = train_test_split(df['new_Text'], df['Cat'], random_state = 0)
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(X_train)
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)
clf = MultinomialNB()
clf.fit(X_train_tfidf, y_train)

MultinomialNB()

In [ ]:

print(clf.predict(count_vect.transform(["I loved the flavor of the bars!!! I really didn't expect them to be as good as they are!!! I have had other nutritional bars that tasted really nasty. I'd buy the Balance Bars again, and plan on doing so. Even my children liked them! The bars gave me energy when I ate them in the late afternoon. They were a real pick-me-up! I'd recommend this product to a friend in a heartbeat. It's nice to know you can buy a nutritional bar that tastes good and seems more like a snack!"])))

['nutrition wellness']


In [ ]:
print(clf.predict(count_vect.transform(['I play with this all the time!!']))) 

['electronics for kids']


In [ ]:
print(clf.predict(count_vect.transform(["If you are one of those people that collects these things, then by all means get it, because it will thrill and amaze you. For you kids, though--basically it'll just sit there, and adults you know will tell you that you shouldn't play with it, because it will be worth something someday. Whatever. Pretty durable if you decide not to listen to them. Not very educational--it doesn't espouse the virtues of peace, other than the cute little peace symbol"])))

['stuffed animals plush']


In [ ]:
#validating the model using a new review 
print(clf.predict(count_vect.transform(['Hi I love these puzzles, I spend all my days playing with them ']))) 

['puzzles']


In [ ]:
X_test_counts = count_vect.transform(X_test)
tfidf_transformer = TfidfTransformer()
X_test_tfidf = tfidf_transformer.fit_transform(X_test_counts)

In [ ]:

import numpy as np 
from sklearn.metrics import accuracy_score

print(X_train_tfidf.shape)
print(X_test_tfidf.shape)
clf.fit(X_train_tfidf,y_train)



(3750, 18195)
(1250, 18195)


MultinomialNB()

In [ ]:
X_train_tfidf

<3750x18195 sparse matrix of type '<class 'numpy.float64'>'
	with 130634 stored elements in Compressed Sparse Row format>

In [ ]:
y_hat = clf.predict(X_test_tfidf)
y_true = y_test
clf_2 = accuracy_score(y_hat,y_true)

In [ ]:
clf_2

0.504

As we see when using Naive Bayes for classification , we get an accuracy acore of 0.504 which is not very good. We can try using other Multi-classification algorithms for the given data and then compare accuracy scores from other algorithms and then compare to see which algorithm would be best for multi- class classification of the given dataset. 


### Q 3.2 Build a Method for Determining a Hierarchy

#### Overview

This exercise is a representation of the challenges we face with our growing topic taxonomy. You are asked to create a method to automatically determine a hierarchy for different products using the “option_2.csv” dataset. There is a list of 5K reviews for several product categories. In this dataset the hierarchy is already provided (each review maps to the child category “Cat1”, and then to the parent category “Cat2”). Separate each review into “Cat1” categories and then given this child category, determine a method to predict the parent category.

#### Hint
A possible approach would be to take the data from each “Cat1” category and create clusters that closely map to each “Cat2” categories.

#### Data
“Text” – A review given by a customer for a purchased item

“Cat1” – A low level child category for each review

“Cat2” – A high level parent category for each review
